pkg/services/ngalert/schedule/schedule_unit_test.go

```diff
@@ -381,9 +385,18 @@ func TestSchedule_ruleRoutine(t *testing.T) {
 	normalStates := []eval.State{eval.Normal, eval.Alerting, eval.Pending}
 	allStates := [...]eval.State{eval.Normal, eval.Alerting, eval.Pending, eval.NoData, eval.Error}
 
+	type normalStatesSetup struct {
+		sch           *schedule
+		instanceStore *state.FakeInstanceStore
+		reg           prometheus.Gatherer
+		rule          *models.AlertRule
+		folderTitle   string
+		expectedTime  time.Time
+	}
+
 	for _, evalState := range normalStates {
-		// TODO rewrite when we are able to mock/fake state manager
-		t.Run(fmt.Sprintf("when rule evaluation happens (evaluation state %s)", evalState), func(t *testing.T) {
+		evalState := evalState // make a local copy to allow parallel tests
```

Comment:
Recommend adding a TODO to remove this when we go to 1.22: https://go.dev/blog/loopvar-preview

Reply:
Fixed!